File: Python/0124-binary-tree-maximum-path-sum.py
# Time:  O(n)
# Space: O(h), h is height of binary tree
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    maxSum = float("-inf")
    def maxPathSum(self, root: TreeNode) -> int:
        self.PathSum(root)
        return self.maxSum


    def PathSum(self, node):
        if node is None:
            return 0

        left = max(0, self.PathSum(node.left))
        right = max(0, self.PathSum(node.right))

        self.maxSum = max(self.maxSum, left + right + node.val)

        return max(left,right) + node.val


# Notes:
# Unlike iterative dfs using stack, here we do work after returning

File: C++/0124-binary-tree-maximum-path-sum.cpp
// Tags: Amazon Tree DFS
// Time:  O(n)
// Space: O(h), h is height of binary tree
class Solution {
public:
    int maxSum = numeric_limits<int>::min();
    int maxPathSum(TreeNode* root) {
        PathSum(root);
        return maxSum;
    }
// PathSum() returns the maximum gain one could have including the node and may
// be one of its subtrees into the path
    int PathSum (TreeNode* node){
        if (node == nullptr)
            return 0;
        // if any branch returns negative path sum, we can skip that branch
        int left = max(0, PathSum(node->left));
        int right = max(0, PathSum(node->right));

        maxSum = max(maxSum, left + right + node->val);

        return max(left,right) + node->val;

    }
};
